Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Forward.h"
+#include "Job.h"
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+#include <AK/Types.h>
+#include <AK/Vector.h>
+#include <LibLine/Editor.h>
+
+using TheExecutionInputType = RefPtr<Shell>;
+
+namespace AST {
+
+struct HighlightMetadata {
+    bool is_first_in_list { true };
+};
+
+struct Position {
+    size_t start_offset { 0 };
+    size_t end_offset { 0 };
+    bool in_range(size_t offset) const { return start_offset <= offset && end_offset >= offset; }
+};
+
+struct Redirection {
+    String path;
+    int fd { 0 };
+    enum {
+        Read,
+        Write,
+        WriteAppend,
+    } direction { Read };
+};
+
+struct Rewiring {
+    int source_fd { 0 };
+    int dest_fd { 0 };
+    enum class Close {
+        None,
+        Source,
+        Destination,
+    } must_be_closed { Close::None };
+};
+
+struct Command {
+    Vector<String> argv;
+    Vector<Redirection> redirections;
+    Vector<Rewiring> rewirings;
+    bool should_be_run_in_background { false };
+    bool is_pipe_source { false };
+};
+
+struct HitTestResult {
+    RefPtr<Node> matching_node;
+    RefPtr<Node> closest_node_with_semantic_meaning; // This is used if matching_node is a bareword
+};
+
+class Value : public RefCounted<Value> {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) = 0;
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType);
+    virtual ~Value();
+    virtual bool is_command() const { return false; }
+    virtual bool is_glob() const { return false; }
+    virtual bool is_job() const { return false; }
+    virtual bool is_list() const { return false; }
+    virtual bool is_string() const { return false; }
+};
+
+class CommandValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType) override;
+    virtual ~CommandValue();
+    virtual bool is_command() const override { return true; }
+    CommandValue(Command command)
```

Comment:
I'd say
```suggestion
    CommandValue(Command&& command)
```
but then I am reminded of the discussion we've had about this the last time, so maybe let's just not do that :smile:

Reply:
Uhh...I'd like to point out that it doesn't really matter here, and...yeah, let's not do that.